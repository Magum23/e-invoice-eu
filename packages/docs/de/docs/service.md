---
main: /en/docs/service.md
---
