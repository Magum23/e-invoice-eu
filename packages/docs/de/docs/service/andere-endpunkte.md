---
main: /en/docs/service/other-endpoints.md
---
