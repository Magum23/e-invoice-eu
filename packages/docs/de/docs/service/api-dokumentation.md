---
main: /en/docs/service/api-documentation.md
---
