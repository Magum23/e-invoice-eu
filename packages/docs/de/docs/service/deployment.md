---
main: /en/docs/service/deployment.md
---
