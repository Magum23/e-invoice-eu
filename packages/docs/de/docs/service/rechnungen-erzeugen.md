---
main: /en/docs/service/create-invoices.md
---
