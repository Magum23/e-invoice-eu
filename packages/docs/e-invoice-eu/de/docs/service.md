---
main: /e-invoice-eu/en/docs/service.md
---
