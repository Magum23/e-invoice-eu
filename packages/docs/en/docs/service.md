---
title: Service
name: service
virtual: 1 
---
