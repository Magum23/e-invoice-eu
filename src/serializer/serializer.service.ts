import { Injectable } from '@nestjs/common';
import { create } from 'xmlbuilder2';
import { ExpandObject } from 'xmlbuilder2/lib/interfaces';

export type SerializerOptions = {
	indent?: string;
	prettyPrint?: boolean;
};

@Injectable()
export class SerializerService {
	serialize(
		data: { [elem: string]: any },
		options: SerializerOptions = {},
	): string {
		const tree = this.convert(data);

		const doc = create({ version: '1.0', encoding: 'utf-8' }).ele(tree);

		return doc.end({
			prettyPrint: options.prettyPrint,
			indent: options.indent,
		});
	}

	private convert(input: { [key: string]: any }): ExpandObject {
		let output: ExpandObject;
		if (Array.isArray(input)) {
			output = [];
			for (let i = 0; i < input.length; ++i) {
				output.push(this.convert(input[i]));
			}
		} else {
			output = {};
			for (const key in input) {
				if (typeof input[key] === 'object') {
					output[key] = this.convert(input[key]);
				} else {
					const [elem, attr] = key.split('@', 2);

					if (typeof attr !== 'undefined') {
						if (!(elem in output)) {
							throw new Error(`orphan attribute ${key}`);
						}

						if (typeof output[elem] === 'string') {
							output[elem] = {
								'#': output[elem],
								[`@${attr}`]: input[key].toString(),
							};
						} else {
							output[elem][`@${attr}`] = input[key].toString();
						}
					} else {
						output[elem] = input[key].toString();
					}
				}
			}
		}

		return output;
	}
}
