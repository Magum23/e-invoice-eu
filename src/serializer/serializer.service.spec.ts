import { Test, TestingModule } from '@nestjs/testing';

import { SerializerOptions, SerializerService } from './serializer.service';

const options: SerializerOptions = {
	prettyPrint: true,
	indent: '\t',
};

describe('SerializerService', () => {
	let service: SerializerService;

	beforeEach(async () => {
		const module: TestingModule = await Test.createTestingModule({
			providers: [SerializerService],
		}).compile();

		service = module.get<SerializerService>(SerializerService);
	});

	it('should be defined', () => {
		expect(service).toBeDefined();
	});

	it('should serialize an object', () => {
		const xml = service.xml('Root', { xmlns: 'urn:mine' }, {}, options);

		expect(xml).toBe(
			'<?xml version="1.0" encoding="utf-8"?>\n<Root xmlns="urn:mine"/>',
		);
	});

	it('should serialize a nested object', () => {
		const data = {
			foo: {
				bar: 'foobar',
			},
			baz: 'bazoo',
		};
		const xml = service.xml('Root', { xmlns: 'urn:mine' }, data, options);

		expect(xml).toMatchSnapshot();
	});

	it('should serialize attributes of text nodes', () => {
		const data = {
			foo: 'bar',
			'foo@baz': 'bazoo',
		};
		const xml = service.xml('attributes', { xmlns: 'urn:mine' }, data, options);

		expect(xml).toMatchSnapshot();
	});

	it('should serialize attributes of object nodes', () => {
		const data = {
			foo: {
				bar: 'baz',
			},
			'foo@baz': 'bazoo',
		};
		const xml = service.xml('attributes', { xmlns: 'urn:mine' }, data, options);

		expect(xml).toMatchSnapshot();
	});

	it('should serialize arrays', () => {
		const data = {
			foo: [
				{
					bar: 1,
					baz: 11,
				},
				{
					bar: 2,
					baz: 22,
				},
				{
					bar: 3,
					baz: 33,
				},
			],
		};
		const xml = service.xml('attributes', { xmlns: 'urn:mine' }, data, options);
		console.warn(xml);

		expect(xml).toMatchSnapshot();
	});
});
