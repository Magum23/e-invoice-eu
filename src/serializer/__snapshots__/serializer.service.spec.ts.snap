// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SerializerService should serialize a nested object 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
<Root xmlns="urn:mine">
	<foo>
		<bar>foobar</bar>
	</foo>
	<baz>bazoo</baz>
</Root>"
`;

exports[`SerializerService should serialize arrays 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
<attributes xmlns="urn:mine">
	<foo>
		<bar>1</bar>
		<baz>11</baz>
	</foo>
	<foo>
		<bar>2</bar>
		<baz>22</baz>
	</foo>
	<foo>
		<bar>3</bar>
		<baz>33</baz>
	</foo>
</attributes>"
`;

exports[`SerializerService should serialize attributes of object nodes 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
<attributes xmlns="urn:mine">
	<foo baz="bazoo">
		<bar>baz</bar>
	</foo>
</attributes>"
`;

exports[`SerializerService should serialize attributes of text nodes 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
<attributes xmlns="urn:mine">
	<foo baz="bazoo">bar</foo>
</attributes>"
`;
